### PR TITLE
Expenses Across Hosts

### PR DESCRIPTION
**Pull request by znarf**
API Part: https://github.com/opencollective/opencollective-api/pull/6899

Project: https://github.com/opencollective/opencollective/issues/4966

### CODE REVIEW

File: components/expenses/ExpenseFormPayeeStep.js [hunk 1]
```diff
@@ -88,12 +87,20 @@ const getPayoutMethodsFromPayee = payee => {
   }
 
   // If the Payee is in the "Collective" family (Collective, Fund, Event, Project)
-  // Then the Account Balance should be its only option
+  // But not the Host itself (Self Hosted)
+  // Then we should add BANK_ACCOUNT and PAYPAL of the Host as an option
   if (payee && AccountTypesWithHost.includes(payee.type) && payee.id !== payee.host?.id) {
-    filteredPms = filteredPms.filter(pm => pm.type === PayoutMethodType.ACCOUNT_BALANCE);
```
Comment (znarf) [inline components/expenses/ExpenseFormPayeeStep.js:93]: That was not needed, it's taken care of later in the process.

File: components/expenses/ExpenseFormPayeeStep.js [hunk 2]
```diff
@@ -88,12 +87,20 @@ const getPayoutMethodsFromPayee = payee => {
...
+    if (!payoutMethods.find(pm => pm.type === PayoutMethodType.BANK_ACCOUNT)) {
+      const hostPayoutMethods = get(payee, 'host.payoutMethods') || EMPTY_ARRAY;
+      const hostBankAccountPayoutMethods = hostPayoutMethods
```
Comment (znarf) [inline components/expenses/ExpenseFormPayeeStep.js:95]: Do we really want PayPal or should we start with Bank Transfer?

If, PayPal is ok, rename variable.

**Review by znarf — COMMENTED** [review]
Check Payment Processor Fees are accurate in Ledger. Back by a snapshot test on API?

File: components/expenses/PayoutMethodSelect.js [hunk 3]
```diff
@@ -195,39 +202,51 @@ class PayoutMethodSelect extends React.Component {
...
+    const payeeIsSameHost = payee && payee.host?.id === this.props.collective.host?.id;
+
+    let pmTypes;
+
+    if (payeeIsSameHost) {
+      pmTypes = [PayoutMethodType.ACCOUNT_BALANCE];
+    } else {
+      pmTypes = payerHostSupportedPayoutMethods
+        // Credit Card (Virtual Card) is generally not a Payout Method acceptable on the Frontend
+        // Should it be completely removed?
```
Comment (znarf) [inline components/expenses/PayoutMethodSelect.js:223]: It's needed for CHARGE expense likely.

File: pages/expense.js [hunk 4]
```diff
@@ -435,20 +435,18 @@ class ExpensePage extends React.Component {
       return [];
     } else {
       const payoutProfiles = [loggedInAccount];
-      for (const node of get(loggedInAccount, 'adminMemberships.nodes', [])) {
+      for (const membership of get(loggedInAccount, 'adminMemberships.nodes', [])) {
```
Comment (znarf) [inline pages/expense.js:438]: This whole block can be shared between the 2 pages.

File: components/expenses/PayoutMethodSelect.js [hunk 5]
```diff
@@ -239,16 +258,58 @@ class PayoutMethodSelect extends React.Component {
...
+    const styledSelectOptions = this.getOptions(payoutMethods, payee);
+    const hasSuitablePayoutMethodOption = styledSelectOptions.find(({ options }) => options.length > 0) ? true : false;
+
+    if (payeeIsCollectiveFamilyType && !payeeIsSameHost) {
+      if (!collective?.host?.isTrustedHost) {
+        return (
+          <MessageBox type="error" mt={2} mb={3} fontSize="12px">
+            <FormattedMessage defaultMessage="This Expense is between different Hosts but the Payer Host is not allowed for this yet." />
+            &nbsp;
+            <FormattedMessage defaultMessage="If it's an issue, contact the Host or Open Collective support." />
```
Comment (Betree) [inline components/expenses/PayoutMethodSelect.js:278]: Why not this:

```suggestion
            <FormattedMessage defaultMessage="This Expense is between different Hosts but the Payer Host is not allowed for this yet. If it's an issue, contact the Host or Open Collective support." />
```

Comment (znarf) [reply components/expenses/PayoutMethodSelect.js:278]: The goal was to reuse the second part elsewhere.

**Pull request merged**